### PR TITLE
Rename `EngineInterface` to `Engine`

### PR DESCRIPTION
**Pull request by nicklan**
Does _a lot_ of renaming:
- `EngineInterface` to `Engine`.
- All `impls` to drop the `Interface` part of the name, and changes any `engine_interface` variables to be just `engine`.
- change the `client` mod name to `engine`
- feature flags: `default-client` -> `default-engine`, `sync-client` -> `sync-engine`
- There were a few leftover places where we still called this `table_client`, this also renames those to `engine`.
- Also this changes the word "interface" to "trait" in some comments where we really meant trait and not interface.

Ensured I got to everything by grepping:
```
$ rg -i EngineInterface
$ rg -i engine_interface
$ rg -i table_client
$ rg -i interface # these are all proper uses of the word interface
LICENSE
45:      separable from, or merely link (or bind by name) to the interfaces of,

acceptance/README.md
6:allowing us to test the exact same client interfaces that any connector

ffi/src/lib.rs
1:/// FFI interface for the delta kernel
3:/// Exposes that an engine needs to call from C/C++ to interface with kernel

ffi/src/scan.rs
42:/// Interface](https://arrow.apache.org/docs/format/CDataInterface.html). This includes the data and
52:/// Interface](https://arrow.apache.org/docs/format/CDataInterface.html). This includes the data and
```

### CODE REVIEW

File: kernel/src/engine/mod.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+//! Provides engine implementation that implement the required traits. These engines can optionally
```
Comment (nicklan) [inline kernel/src/engine/mod.rs:1]: This file is untouched, logic wise, from the old `client/mod.rs`. But with a change in the mod comment plus the feature flag name changes, github isn't smart enough to realize what happened and thinks it's a new file and that the old one was removed.

**Review by scovich — APPROVED** [review]
LGTM

Rebasing other pull requests won't be fun, but I love how many places multi-line statements became single-line due to the shorter names.

**Review by roeap — APPROVED** [review]
^^ what ryan said :)

**Comment by nicklan** [comment]
Thanks all. Sorry for the extra rebasing work!

**Pull request merged**